### PR TITLE
add numeric comparison

### PR DESCRIPTION
**Pull request by grusy**
*Issue #14*

*Description of changes:* 
add numeric comparison operators

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cfn-guard/README.md [hunk 1]
```diff
@@ -89,6 +90,8 @@ The available operations are:
 
 * `==` - Equal
 * `!=` - Not Equal
+* `<=` - Smaller As
```
Comment (rjlohan) [inline cfn-guard/README.md:93]: I think it would be more familiar to use "Less Than" and "Greater Than" for these labels.

Comment (rjlohan) [reply cfn-guard/README.md:93]: The token also suggests that the evaluation would be "Less Than or Equal To" which is not how your rule evaluates. We might want to clear that up. :+1:

File: cfn-guard/src/parser.rs [hunk 2]
```diff
@@ -156,6 +156,8 @@ fn destructure_rule(rule_text: &str, cfn_resources: &HashMap<String, Value>) ->
                     match &caps["operator"] {
                         "==" => OpCode::Require,
                         "!=" => OpCode::RequireNot,
+                        "<=" => OpCode::SmallerAs,
```
Comment (nathanataws) [inline cfn-guard/src/parser.rs:159]: This is a great start!  I think we need to smooth out the ergonomics of the operators a bit, though.  

I think there are two subtle but important changes:
1.  The numeric operator should be on the left side of the `=` in both cases or it starts to look like an arrow, and
2.  The rule operator looks like it's saying "lesser/greater than or equal to" but the actual check is doesn't include equality.  I think we should expand this to two more operators & operations:

< Less Than
> Greater Than
<= Less than or equal to
>= Greater than or equal to

File: cfn-guard/src/lib.rs [hunk 3]
```diff
@@ -377,6 +377,70 @@ fn apply_rule_operation(
                 None
             }
         }
+        enums::OpCode::SmallerAs => {
```
Comment (nathanataws) [inline cfn-guard/src/lib.rs:380]: I think we should have 4 total operations (see my other comment):

< Less Than
> Greater Than
<= Less than or Equal To
>= Greater than or Equal to

Comment (grusy) [reply cfn-guard/src/lib.rs]: Added the two additional comparisons and fixed the other issues.

Comment (nathanataws) [reply cfn-guard/src/lib.rs]: Excellent.  Will review this week.  Many thanks for the quick turnaround!

File: cfn-guard/src/lib.rs [hunk 4]
```diff
@@ -377,6 +377,136 @@ fn apply_rule_operation(
...
+                _ => {
+                    error!("REQUIRE rule type that doesn't match RValueType of Regex, Variable or Value");
+                    None
+                }
+            }
+        }
+        enums::OpCode::GreaterThanOrEqualTo => {
+            match rule.rule_vtype {
+                enums::RValueType::Value | enums::RValueType::Variable => {
+                    if util::format_value(&val).parse::<f32>().unwrap() < util::strip_ws_nl(rule_val.to_string()).parse::<f32>().unwrap() {
```
Comment (nathanataws) [inline cfn-guard/src/lib.rs:481]: I think the operators for the LessThan/GreaterThan & LessThanOrEqualTo/GreaterThanOrEqualTo are reversed

Comment (nathanataws) [reply cfn-guard/src/lib.rs:481]: I read this in reverse.  I think all four of these would read more naturally if the operators were the same as the OpCode for the match arm.

For instance, `LessThanOrEqualTo` should use the `<=` rust numeric operator and have the first half of the `if` equal a pass.  In other words, fail it if it fails rust's numeric comparison. 

Ditto for all 4 cases.

File: cfn-guard/src/lib.rs [hunk 5]
```diff
@@ -377,6 +377,136 @@ fn apply_rule_operation(
                 None
             }
         }
+        enums::OpCode::LessThan => {
+            match rule.rule_vtype {
+                enums::RValueType::Value | enums::RValueType::Variable => {
+                    if util::format_value(&val).parse::<f32>().unwrap() > util::strip_ws_nl(rule_val.to_string()).parse::<f32>().unwrap() ||
```
Comment (nathanataws) [inline cfn-guard/src/lib.rs:383]: Similar to my comment on GreaterThanOrEqaulTo, I would just use the rust `<` numeric operator here and flip the pass/fail case.

File: cfn-guard/src/lib.rs [hunk 6]
```diff
@@ -377,6 +377,136 @@ fn apply_rule_operation(
...
+                _ => {
+                    error!("REQUIRE rule type that doesn't match RValueType of Regex, Variable or Value");
+                    None
+                }
+            }
+        }
+        enums::OpCode::GreaterThan => {
+            match rule.rule_vtype {
+                enums::RValueType::Value | enums::RValueType::Variable => {
+                    if util::format_value(&val).parse::<f32>().unwrap() < util::strip_ws_nl(rule_val.to_string()).parse::<f32>().unwrap() ||
```
Comment (nathanataws) [inline cfn-guard/src/lib.rs:416]: Use `>` here and flip the pass-fail

File: cfn-guard/src/lib.rs [hunk 7]
```diff
@@ -377,6 +377,136 @@ fn apply_rule_operation(
...
+                _ => {
+                    error!("REQUIRE rule type that doesn't match RValueType of Regex, Variable or Value");
+                    None
+                }
+            }
+        }
+        enums::OpCode::LessThanOrEqualTo => {
+            match rule.rule_vtype {
+                enums::RValueType::Value | enums::RValueType::Variable => {
+                    if util::format_value(&val).parse::<f32>().unwrap() > util::strip_ws_nl(rule_val.to_string()).parse::<f32>().unwrap() {
```
Comment (nathanataws) [inline cfn-guard/src/lib.rs:449]: Use `<=` here and flip the pass-fail

File: cfn-guard/tests/functional.rs [hunk 8]
```diff
@@ -814,6 +814,45 @@ AWS::EC2::Volume Size == 101 |OR| AWS::EC2::Volume Size == 99"#,
         );
     }
 
+    #[test]
+    fn test_or_should_pass_numeric() {
```
Comment (nathanataws) [inline cfn-guard/tests/functional.rs:818]: I believe this is confirming that it fails correctly.  If so, can you update the test name to indicate that?

Please update the test with multiple smaller internal `assert_eq!` to check that it fails correctly for each individual operator (the `|OR|` here could obscure a bad path somewhere).

Also, please add tests for:
1.  Showing that each of the new operators passes correctly
2.  That it handles a non-numeric value being used in a numeric comparison as a runtime error that exits with a `1` and returns a message with the text from the rule that failed.

**Comment by nathanataws** [comment]
Also, can you change the branch you're PR'ing into to `release-0.5.1`.  Thanks!

File: cfn-guard/src/lib.rs (hunk 4 above)
Comment (grusy) [reply cfn-guard/src/lib.rs:481]: I was following the structure of the existing code. OpCode::Require is “!=“. I think, for consistency all comparisons should follow a similar structure.

Comment (nathanataws) [reply cfn-guard/src/lib.rs:481]: That's a fair point.  I think what we're seeing is that that initial decision to have the fail case be the first half of the `if-else` starts to get more and more confusing as we add more operators like these.

Let's do this:  You flip yours around here and I'll flip the other ones to make them consistent.

**Comment by nathanataws** [comment]
@grusy - My apologies.  I had to ship the 0.5.1 release and delete its branch.  Let's open this up and mark it for the `release-0.5.2` branch that I'm going to push up in a bit.  For now, I'll just point this back to `master`.  

Also:  I flipped the "pass/fails" the way we discussed above in the 0.5.2 branch I'm working on.

**Comment by nathanataws** [comment]
Okay.  `release-0.5.2` pushed with the pass-fail flip.

**Comment by nathanataws** [comment]
Okay.   I figured out how to do this in the new GH interface.  I had to go back to the PR when I merged 0.5.1 into master.  From there I could restore the release-0.5.1 branch and reopen this PR.  From there I was able to change the target branch.  Wow.  That is suboptimal.

**Comment by nathanataws** [comment]
So rebase on top of `release-0.5.2` and we can get this into the next release.

**Comment by nathanataws** [comment]
Good morning! I think we ending up working on this at the exact same time.  I'm going to pull your changes into this new merge branch and then I'll merge both to `release-0.5.2`

**Pull request merged**